File: backend/internal/domain/pihole_client.go
package domain

import "time"

type PiholeClient struct {
	Id           int64
	IP           string
	Name         string
	Comment      *string
	Groups       []int
	DateAdded    time.Time
	DateModified time.Time
}

type PiholeClientSet struct {
	Clients []PiholeClient
}

type UpdatePiholeClientCommand struct {
	Id      int64
	Groups  []int
	Comment *string
}

type RemovePiholeClientCommand struct {
	Id int64
}

File: frontend/src/types/pihole_client.ts
import type { PiholeNodeRef } from './pihole';

export type PiholeClient = {
	id: number;
	ip: string;
	name: string;
	comment: string | null;
	groups: number[];
	dateAdded: string;
	dateModified: string;
};

// Per-node result

type ListClientsNode = {
	node: PiholeNodeRef;
	clients: PiholeClient[];
	error?: string;
};

type ListClientsSummary = {
	totalNodes: number;
	okNodes: number;
	errorNodes: number;
};

export type ListClientsResponse = {
	summary: ListClientsSummary;
	nodes: Record<string, ListClientsNode>;
};

type ClientMutateNode = {
	node: PiholeNodeRef;
	clients: PiholeClient[];
	error?: string;
};

export type ClientMutateResponse = {
	nodes: Record<string, ClientMutateNode>;
};

type RemoveClientNode = {
	node: PiholeNodeRef;
	removed: boolean;
	error?: string;
};

type RemoveClientSummary = {
	total: number;
	removed: number;
	failed: number;
};

export type RemoveClientResponse = {
	summary: RemoveClientSummary;
	nodes: Record<string, RemoveClientNode>;
};
